main.js

```diff
@@ -282,44 +314,57 @@ loops = {
 	},
 
 	upgrade: () => {
+		// leave if all firewalls are upgraded to max
+		if (!vars.fireWall[3].needUpgrade)
+			return;
+		const i = getRandomInt(1,3);
+		// if this fireWall is already fully upgraded, get an other random firewall.
+		if (!vars.fireWall[i - 1].needUpgrade)
+			vars.loops.upgrade();
 		vars.balance = parseInt($("#window-my-coinamount").text());
-		// if the back button is visible, we're on a page, let's back out
+		// if the back button is visible, we're on a page, let's back out and hide the firewall warning.
 		if ($("#window-firewall-pagebutton").is(":visible") === true) {
+			$("#tutorial-firewall").css("display","none");
 			$("#window-firewall-pagebutton").click();
 		}
+
 		// just get a random port, because who cares
```

Comment:
This comment is out of place, looks like it should be moved up to above line 320